### DISCUSSION TITLE
Integrate with TanStack Router

### DISCUSSION THREAD

**Discussion by dai-shi** [wakujs/waku#272] [Ideas]
Waku's minimal API is designed to be router agnostic.
It would be nice if integration with TanStack Router is possible.
Would anyone like to give a challenge? First, investigate the feasibility.

**Comment by dai-shi** [comment]
backlink: https://twitter.com/tannerlinsley/status/1735432759768658119

**Comment by Aslemammad** [comment]
I'd like to take this.

**Reply by dai-shi** [reply]
Looking forward to it!

My first trial was #118, but it's too rough.

**Comment by ojj1123** [comment]
@dai-shi 

> Waku's minimal API is designed to be router agnostic.

Why is it intended to be router-agnostic?
Could I understand as the intend using just a various of router solution's features(such as `loader`, `cache`, etc) w/o implementing this features on Waku?

**Reply by dai-shi** [reply]
Waku started as an experimental project, and while we now focus on making it a production-level RSC framework, we want others to experiment RSC and come up with new solutions. So, the minimal API is for such library authors.

We'll providing an official router `waku/router`. #293 
It's built on top of the minimal API. So, Waku's core doesn't have router features.

**Comment by LizAinslie** [comment]
I have been hacking at this for a little bit now and i've managed to put together a small sample of Waku + Tanstack Router on the latest versions of each here: https://github.com/LizAinslie/waku-tanstack-poc

It works in dev mode, however there are some really obscure errors produced when building and starting for production

**Reply by dai-shi** [reply]
I'm thinking about changing our minimal api for a following release. So, it would be nice to learn from tanstack router integration. Taking a closer look. 👀

**Reply by dai-shi** [reply]
So, while I'm not sure if I understand every details, it looks to me that TSR is a client library for SPA (optionally with multi-page SSR), so the integration with Waku mostly means as a SPA.
The benefit of using Waku with TSR will be server actions, and possibly rsc-based data fetching (which is still to be explored.)

What concerns me is TSR now pushes file-based routing https://tanstack.com/router/latest/docs/framework/react/guide/file-based-routing, which generally conflicts with Waku's fs router. The TSR's config-based router should be feasible to integrate with Waku, but I wonder if it's appealing to TSR users.

**Reply by tylersayshi** [reply]
@dai-shi What features would you like to get from TSR and which would you like to leave out?

It feels like from reading above that TSR is a bit too opinionated to fit fully in a waku project, but some of the features could be really helpful. 

I really like the Link's and typescript support for routes that TSR has implemented and I'd imagine something like that could just be a feature that waku/router replicates (I'd be happy to help with adding something like that).

**Reply by dai-shi** [reply]
My motivation is to prove that Waku core is router agnostic.
I agree TSR is too opinionated to fit well with Waku.

> typescript support for routes

Yeah, they are nice.

> (I'd be happy to help with adding something like that)

Yeah, that would be helpful. Feel free to work on it. One note is that I'm planning to redesign the minimal api `defineEntries` and router integration, and especially around buildData and skipList. So, we shouldn't change the core part yet. If the modification is in the higher level, it should be fine.

**Comment by Aslemammad** [comment]
That's amazing, I'd like to use this as an inspiration! Do you have discord?

**Comment by dai-shi** [comment]
#1165 is _one_ of the answers.

**Comment by nyomansunima** [comment]
Cool idea
